src/s3.rs

```diff
@@ -0,0 +1,211 @@
+                .client
+                .list_objects_v2()
+                .bucket(bucket)
+                .set_continuation_token(continuation_token)
+                .prefix(prefix)
+                .send()
+                .await;
+
+            match result {
+                Ok(data) => {
+                    continuation_token = data.continuation_token().map(String::from);
+                    for content in data.contents().unwrap() {
+                        res.push(OutputS3ObjectMetadata {
+                            key: String::from(
+                                content.key().expect("key value should exist on s3 object"),
+                            ),
+                            last_modified: content
+                                .last_modified()
+                                .expect("last_modified should exist on s3 object")
+                                .to_millis()
+                                .unwrap(),
+                            size: content.size(),
+                        });
+                    }
+                }
+                Err(e) => {
+                    let msg = format!("error while s3 list_files {}/{}; err={}", bucket, prefix, e);
+                    tracing::error!(msg);
+                    return Err(AwsError::new(msg.as_str()));
+                }
+            }
+        }
+        Ok(res)
+    }
+
+    pub async fn copy_object(
+        &self,
+        bucket_to_copy: &str,
+        file_to_copy: &str,
+        destination_bucket: &str,
+        destination_file: &str,
+    ) -> Result<(), AwsError> {
+        let source_file = format!("{}/{}", bucket_to_copy, file_to_copy);
+        let result = self
+            .client
+            .copy_object()
+            .copy_source(source_file.clone())
+            .bucket(destination_bucket)
+            .key(destination_file)
+            .send()
+            .await;
+
+        match result {
+            Ok(_) => Ok(()),
+            Err(e) => {
+                let error_message = format!(
+                    "error while s3 copy_object {} to {}/{}; err={}",
+                    source_file, destination_bucket, destination_file, e
+                );
+                tracing::error!(error_message);
+                Err(AwsError::new(error_message.as_str()))
+            }
+        }
+    }
+
+    pub async fn put_object(&self, data: Vec<u8>, bucket: &str, key: &str) -> Result<(), AwsError> {
+        let result = self
+            .client
+            .put_object()
+            .bucket(bucket)
+            .key(key)
+            .body(ByteStream::from(data))
+            .send()
+            .await;
+
+        match result {
+            Ok(_) => Ok(()),
+            Err(e) => {
+                let error_message =
+                    format!("error while s3 put_object {}/{}; err={}", bucket, key, e);
+                tracing::error!(error_message);
+                Err(AwsError::new(&*error_message))
```

Comment:
Suggestion:
```suggestion
                Err(AwsError::new(&error_message))
```

Moreover, I see that `AwsError`s are built using either `&error_message` or `error_message.as_str()`.
Let's be consistent and use the same everywhere (I don't have any preference about which one we should use).